mesos-client/src/main/scala/mesosphere/mesos/client/MesosClient.scala

```diff
@@ -0,0 +1,514 @@
+    */
+
+  /** This is the first step in the communication process between the scheduler and the master. This is also to be
+    * considered as subscription to the “/scheduler” event stream. To subscribe with the master, the scheduler sends
+    * an HTTP POST with a SUBSCRIBE message including the required FrameworkInfo. Note that if
+    * `subscribe.framework_info.id` is not set, master considers the scheduler as a new one and subscribes it by
+    * assigning it a FrameworkID. The HTTP response is a stream in RecordIO format; the event stream begins with a
+    * SUBSCRIBED event.
+    *
+    * Note: this method is used by mesos client to establish connection to mesos master and is not supposed to be called
+    * directly by the framework.
+    *
+    * http://mesos.apache.org/documentation/latest/scheduler-http-api/#subscribe-1
+    */
+  protected def subscribe(frameworkInfo: FrameworkInfo): Call = {
+    Call(
+      frameworkId = frameworkInfo.id,
+      subscribe = Some(Call.Subscribe(frameworkInfo)),
+      `type` = Some(Call.Type.SUBSCRIBE)
+    )
+  }
+
+  /** Sent by the scheduler when it wants to tear itself down. When Mesos receives this request it will
+    * shut down all executors (and consequently kill tasks). It then removes the framework and closes all
+    * open connections from this scheduler to the Master.
+    *
+    * http://mesos.apache.org/documentation/latest/scheduler-http-api/#teardown
+    */
+  def teardown(): Call = {
+    Call(
+      `type` = Some(Call.Type.TEARDOWN)
+    )
+  }
+
+  /** Sent by the scheduler when it accepts offer(s) sent by the master. The ACCEPT request includes the type
+    * of operations (e.g., launch task, launch task group, reserve resources, create volumes) that the scheduler
+    * wants to perform on the offers. Note that until the scheduler replies (accepts or declines) to an offer,
+    * the offer’s resources are considered allocated to the offer’s role and to the framework.
+    *
+    * http://mesos.apache.org/documentation/latest/scheduler-http-api/#accept
+    */
+  def accept(accepts: Accept): Call = {
+    Call(
+      `type` = Some(Call.Type.ACCEPT),
+      accept = Some(accepts))
+  }
+
+  /** Sent by the scheduler to explicitly decline offer(s) received. Note that this is same as sending an ACCEPT
+    * call with no operations.
+    *
+    * http://mesos.apache.org/documentation/latest/scheduler-http-api/#decline
+    */
+  def decline(offerIds: Seq[OfferID], filters: Option[Filters] = None): Call = {
+    Call(
+      `type` = Some(Call.Type.DECLINE),
+      decline = Some(Decline(offerIds = offerIds, filters = filters))
+    )
+  }
+
+  /** Sent by the scheduler to remove any/all filters that it has previously set via ACCEPT or DECLINE calls.
+    *
+    * http://mesos.apache.org/documentation/latest/scheduler-http-api/#revive
+    */
+  def revive(roles: Option[String] = None): Call = {
```

Comment:
`roles: Option[String]`? Should be a collection?

Reply:
No, revive is only defined for one role in v1. `roles` (plural) is whats confusing here. Fixed